File: src/migrations.py
"""Migration file handling and version tracking."""

import re
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import List, Optional


@dataclass
class MigrationScript:
    """Represents a single migration script."""
    
    version: str
    name: str
    up_file: Path
    down_file: Optional[Path] = None
    applied_at: Optional[datetime] = None
    checksum: str = ""
    
    @classmethod
    def from_file(cls, file_path: Path) -> Optional["MigrationScript"]:
        """Parse migration script from filename.
        
        Expected format: V{version}__{name}.sql or V{version}__{name}_down.sql
        Example: V001__create_users_table.sql, V001__create_users_table_down.sql
        """
        pattern = r"^V(\d+)__(.+?)(?:_down)?\.sql$"
        match = re.match(pattern, file_path.name)
        
        if not match:
            return None
        
        version = match.group(1).zfill(3)  # Normalize to 3 digits
        name = match.group(2)
        is_down = "_down" in file_path.name
        
        return cls(
            version=version,
            name=name,
            up_file=file_path if not is_down else None,
            down_file=file_path if is_down else None,
        )
    
    def compute_checksum(self) -> str:
        """Compute SHA256 checksum of the migration script."""
        import hashlib
        
        if not self.up_file or not self.up_file.exists():
            return ""
        
        content = self.up_file.read_text(encoding="utf-8")
        return hashlib.sha256(content.encode()).hexdigest()
    
    @property
    def full_name(self) -> str:
        """Return full migration name with version."""
        return f"V{self.version}__{self.name}"


@dataclass
class MigrationBatch:
    """Represents a batch of migrations to be applied."""
    
    scripts: List[MigrationScript] = field(default_factory=list)
    start_time: Optional[datetime] = None
    end_time: Optional[datetime] = None
    status: str = "pending"  # pending, running, completed, failed, rolled_back
    
    def add_script(self, script: MigrationScript):
        """Add a migration script to the batch."""
        self.scripts.append(script)
        self.scripts.sort(key=lambda s: s.version)
    
    @property
    def versions(self) -> List[str]:
        """Return list of versions in the batch."""
        return [s.version for s in self.scripts]


class MigrationScanner:
    """Scans directory for migration scripts."""
    
    def __init__(self, migrations_dir: Path):
        self.migrations_dir = migrations_dir
    
    def scan(self) -> List[MigrationScript]:
        """Scan directory and return all migration scripts."""
        if not self.migrations_dir.exists():
            return []
        
        migrations = {}
        
        for file_path in self.migrations_dir.glob("*.sql"):
            script = MigrationScript.from_file(file_path)
            if script:
                version = script.version
                
                if version not in migrations:
                    migrations[version] = script
                else:
                    existing = migrations[version]
                    if script.down_file:
                        existing.down_file = script.down_file
                    elif script.up_file:
                        existing.up_file = script.up_file
        
        result = list(migrations.values())
        result.sort(key=lambda s: s.version)
        return result
    
    def validate_scripts(self, scripts: List[MigrationScript]) -> List[str]:
        """Validate migration scripts and return list of errors."""
        errors = []
        
        for script in scripts:
            if not script.up_file or not script.up_file.exists():
                errors.append(f"Missing UP script for version {script.version}")
            elif not script.up_file.read_text(encoding="utf-8").strip():
                errors.append(f"Empty UP script for version {script.version}")
        
        # Check for gaps in version sequence (optional warning)
        if scripts:
            versions = [int(s.version) for s in scripts]
            expected = list(range(min(versions), max(versions) + 1))
            missing = set(expected) - set(versions)
            if missing:
                errors.append(f"Warning: Missing versions: {sorted(missing)}")
        
        return errors

File: dotnet/SqlMigrationBlazor/Models/Migrations.cs
using System.Security.Cryptography;
using System.Text;

namespace SqlMigrationBlazor.Models;

/// <summary>
/// Represents a single migration script.
/// </summary>
public class MigrationScript
{
    public string Version { get; set; } = string.Empty;
    public string Name { get; set; } = string.Empty;
    public FileInfo? UpFile { get; set; }
    public FileInfo? DownFile { get; set; }
    public DateTime? AppliedAt { get; set; }
    public string Checksum { get; set; } = string.Empty;

    /// <summary>
    /// Parse migration script from filename.
    /// Expected format: V{version}__{name}.sql or V{version}__{name}_down.sql
    /// Example: V001__create_users_table.sql, V001__create_users_table_down.sql
    /// </summary>
    public static MigrationScript? FromFile(FileInfo filePath)
    {
        var pattern = @"^V(\d+)__(.+?)(?:_down)?\.sql$";
        var match = System.Text.RegularExpressions.Regex.Match(filePath.Name, pattern);

        if (!match.Success)
            return null;

        var version = match.Groups[1].Value.PadLeft(3, '0');
        var name = match.Groups[2].Value;
        var isDown = filePath.Name.Contains("_down");

        return new MigrationScript
        {
            Version = version,
            Name = name,
            UpFile = !isDown ? filePath : null,
            DownFile = isDown ? filePath : null
        };
    }

    /// <summary>
    /// Compute SHA256 checksum of the migration script.
    /// </summary>
    public string ComputeChecksum()
    {
        if (UpFile == null || !UpFile.Exists)
            return string.Empty;

        var content = File.ReadAllText(UpFile.FullName, Encoding.UTF8);
        using var sha256 = SHA256.Create();
        var hash = sha256.ComputeHash(Encoding.UTF8.GetBytes(content));
        return BitConverter.ToString(hash).Replace("-", "").ToLowerInvariant();
    }

    /// <summary>
    /// Return full migration name with version.
    /// </summary>
    public string FullName => $"V{Version}__{Name}";
}

/// <summary>
/// Represents a batch of migrations to be applied.
/// </summary>
public class MigrationBatch
{
    public List<MigrationScript> Scripts { get; set; } = new();
    public DateTime? StartTime { get; set; }
    public DateTime? EndTime { get; set; }
    public string Status { get; set; } = "pending"; // pending, running, completed, failed, rolled_back

    public void AddScript(MigrationScript script)
    {
        Scripts.Add(script);
        Scripts = Scripts.OrderBy(s => s.Version).ToList();
    }

    public List<string> Versions => Scripts.Select(s => s.Version).ToList();
}

/// <summary>
/// Scans directory for migration scripts.
/// </summary>
public class MigrationScanner
{
    private readonly DirectoryInfo _migrationsDir;

    public MigrationScanner(string migrationsDir)
    {
        _migrationsDir = new DirectoryInfo(migrationsDir);
    }

    public List<MigrationScript> Scan()
    {
        if (!_migrationsDir.Exists)
            return new List<MigrationScript>();

        var migrations = new Dictionary<string, MigrationScript>();

        foreach (var file in _migrationsDir.GetFiles("*.sql"))
        {
            var script = MigrationScript.FromFile(file);
            if (script != null)
            {
                var version = script.Version;

                if (!migrations.ContainsKey(version))
                {
                    migrations[version] = script;
                }
                else
                {
                    var existing = migrations[version];
                    if (script.DownFile != null)
                        existing.DownFile = script.DownFile;
                    else if (script.UpFile != null)
                        existing.UpFile = script.UpFile;
                }
            }
        }

        var result = migrations.Values.ToList();
        result.Sort((a, b) => string.Compare(a.Version, b.Version, StringComparison.Ordinal));
        return result;
    }

    public List<string> ValidateScripts(List<MigrationScript> scripts)
    {
        var errors = new List<string>();

        foreach (var script in scripts)
        {
            if (script.UpFile == null || !script.UpFile.Exists)
            {
                errors.Add($"Missing UP script for version {script.Version}");
            }
            else if (string.IsNullOrWhiteSpace(File.ReadAllText(script.UpFile.FullName, Encoding.UTF8)))
            {
                errors.Add($"Empty UP script for version {script.Version}");
            }
        }

        // Check for gaps in version sequence (optional warning)
        if (scripts.Count > 0)
        {
            var versions = scripts.Select(s => int.Parse(s.Version)).ToList();
            var expected = Enumerable.Range(versions.Min(), versions.Max() - versions.Min() + 1).ToList();
            var missing = expected.Except(versions).ToList();
            if (missing.Count > 0)
            {
                errors.Add($"Warning: Missing versions: {string.Join(", ", missing)}");
            }
        }

        return errors;
    }
}
